### DISCUSSION TITLE
Place delayed jobs at the end of the waiting list

### DISCUSSION THREAD

**Discussion by kietheros** [OptimalBits/bull#2130] [Q&A]
## Description
I have a queue with rate limiter

    limiter: {
        max: 1,
        duration: 4000,
        groupKey: 'sourceId'
    }

    // add 20 jobs with sourceId 'Z'
    for (let i = 0; i < 20; i++) {
        await queue.add({
            sourceId: 'Z'
        }, {
            jobId: `Z:${i+1}`
        });
    }
    
    // add 100 jobs with sourceId 'A', 'B', 'C', 'D', 'E'
    for (let i = 0; i < 20; i++) {
        let sourceIds = ['A', 'B', 'C', 'D', 'E'];
        for (let j = 0; j < 5; j++) {
            const sourceId = sourceIds[j];
            await queue.add({
                    sourceId: sourceId
                }, {
                    jobId: `${sourceId}:${i+1}`,
                });
        }
    }
    
    queue.process(1, async (job) => {
        logger.debug('Start job', job.id);
        await sleep(2000); // sleep 2s
        logger.debug('Finish job', job.id);
    });

Because of rate limiter, some jobs of source Z are moved into delayed queue. After amount of time,  jobs of source `Z` are placed at the beginning of waiting list. The beginning of waiting list here is the left side, so 'Z' jobs will not be processed as soon as the worker is idle.

I have read docs
>  a delayed job will not be processed directly, instead it will be placed **at the beginning of the waiting list** and processed as soon as a worker is idle.

I wonder that these jobs have to be placed at the *end of the waiting list (right side) to be processed as soon as a worker is idle*. ???
Could anyone explain me about this problem ? Thanks !
## Bull version
3.27.0

**Comment by kietheros** [comment]
@manast 
Might I take a moment of your time to ask a small question !!!

In above test case, why delayed jobs are placed at the left of waiting list, and will not be processed immediately, because workers pop jobs from the right of waiting list.

However, docs says:
> a delayed job will not be processed directly, instead it will be placed at the **beginning of the waiting list and processed as soon as** a worker is idle.

**Comment by manast** [comment]
When a delayed job is moved back to the wait list it is placed to the head of the list, i.e. it will be next to be processed... of course if other delayed jobs are moved to the wait list after the first one then they will move the first one back in the queue.

**Reply by kietheros** [reply]
I don't understand. I think delayed jobs need to be moved at the right of waiting list, then idling workers just can get them. If you move them into the left of waiting list, they will be processed after.

Example:
1. Add 20 jobs Z
2. After that, add 100 other jobs

Z jobs hit rate limit, they are move to delayed queue. After amount of time, they should be moved into the right of waiting list to be processed as soon as possible. Because jobs 'Z' are added before others jobs. Could Bull have a option to do this !

**Reply by manast** [reply]
If the queue is like this: ```[ j5, j4, j3, j2, j1, j0]``` and a delayed job expires then it is moved to the head of the queue:  ```[ j5, j4, j3, j2, j1, j0, d0]``` and will be processed by a worker if there are idle workers. But if there are no idle workers and more delayed jobs expire then they will be pushing d0 into the queue: ```[ j5, j4, j3, j2, j1, j0, d0, d1, d2, ...]```

**Reply by kietheros** [reply]
In my test, the results are:
- First, wait list:  `[j100,.., j3, j2, j1, j0]`. `j0, j1` will hit rate limit,  `j2, j3,...,j100` are added and arranged to never hit rate limit.
- When `j0, j1` hit rate limit, they moved to delayed queue `[d1, d0]`. 
- After amount of time, delayed jobs moved back into wait list, `[d1, d0, j100, j99,... j4]`

**Reply by kietheros** [reply]
@manast 
I think this is a bug. You can use the above test case to reproduce this bug.

**Reply by manast** [reply]
What is the bug?

**Reply by kietheros** [reply]
Your expected results: [ j5, j4, j3, j2, j1, j0, d0, d1, d2, ...]
My test results: [d1, d0, j100, j99,... j4]

The bug is delays jobs were moved to the left of wait list instead of the right as in docs.
Test code to reproduce the issue in the first comment.

**Reply by manast** [reply]
Create an issue with minimal test code that reproduces it and I can look into it.

**Reply by manast** [reply]
The code above does not reproduce what you state btw... You should not use the rate limiter in order to reproduce the bug that you are stating.
